### PR TITLE
Switch to cozydb

### PR DESCRIPTION
**Pull request by aenario**
Also include #134 #154 #77

### CODE REVIEW

**Comment by clochix** [comment]
Thanks @aenario :-)
@frankrousseau I probably won't find time to review this this week-end. Would you mind have a look at it so we can merge ASAP ? Thanks !

**Comment by frankrousseau** [comment]
Ok @clochix , I'll do it.

File: client/app/components/toolbox_actions.coffee [hunk 1]
```diff
@@ -56,7 +54,7 @@ module.exports = ToolboxActions = React.createClass
                 if @props.message?
                     li role: 'presentation',
                         a
-                            href: "raw/#{@props.mailboxID}/#{messageRawId}"
+                            href: "raw/#{@props.message.get('id')}"
```
Comment (frankrousseau) [inline client/app/components/toolbox_actions.coffee]: Is that related to CozyDB. Are you sure about what you are doing?

File: server/config.coffee [hunk 2]
```diff
@@ -34,7 +34,7 @@ config =
     ]
 
     plugins: [
-        MODEL_MODULE
+        'cozydb'
```
Comment (frankrousseau) [inline server/config.coffee]: Does it keep the pouchb compatibility?

File: server/controllers/messages.coffee [hunk 3]
```diff
@@ -185,12 +186,17 @@ contentToBuffer = (req, attachment, callback) ->
...
-            return callback err if err
-            callback null, Buffer.concat buffers
+        chunks = []
+        bufferer = new stream.Writable
+        bufferer._write = (chunk, enc, next) ->
+            chunks.push(chunk)
+            next()
+        bufferer.end = ->
+            callback null, Buffer.concat chunks
+        fileStream.pipe bufferer
```
Comment (frankrousseau) [inline server/controllers/messages.coffee]: Why do you need that? Does it mean you will keep the full file in memory?

File: server/controllers/messages.coffee [hunk 4]
```diff
@@ -350,23 +358,43 @@ module.exports.send = (req, res, next) ->
     async.series steps, (err) ->
         return next err if err
         return next new Error('Server error') unless jdbMessage
-        res.send 200, jdbMessage.toClientObject()
+        out = jdbMessage.toClientObject()
+        out.isDraft = isDraft
+        res.send 200, out
```
Comment (frankrousseau) [inline server/controllers/messages.coffee]: Use the new Express API:

```coffeescript
res.status(200).send out
```

By the way, if it's a 200 code, you don't need to mention it.

File: server/controllers/messages.coffee [hunk 5]
```diff
@@ -350,23 +358,43 @@ module.exports.send = (req, res, next) ->
...
 
 
 module.exports.fetchConversation = (req, res, next) ->
```
Comment (frankrousseau) [inline server/controllers/messages.coffee]: Comments?

File: server/controllers/messages.coffee [hunk 6]
```diff
@@ -350,23 +358,43 @@ module.exports.send = (req, res, next) ->
...
+            next NotFound "conversation##{conversationID}"
+        else
+            log.debug "conversation##{conversationID}", messages.length
+            req.conversation = messages
+            next()
 
 module.exports.conversationGet = (req, res, next) ->
     res.send 200, req.conversation.map (msg) -> msg.toClientObject()
 
 module.exports.conversationDelete = (req, res, next) ->
```
Comment (frankrousseau) [inline server/controllers/messages.coffee]: Comments?

File: server/controllers/messages.coffee [hunk 7]
```diff
@@ -350,23 +358,43 @@ module.exports.send = (req, res, next) ->
...
+    accountID = req.conversation[0].accountID
+
+    Account.find accountID, (err, account) ->
+        return next err if err
+        return next new NotFound "Account##{accountID}" unless account
 
-    # @TODO : Delete Conversation
-    res.send 200, []
+        console.log account
```
Comment (frankrousseau) [inline server/controllers/messages.coffee]: Dirty logging

File: server/imap/connection.coffee [hunk 8]
```diff
@@ -204,6 +203,13 @@ Imap::multicopy = (uid, paths, callback) ->
         @copy uid, path, cb
     , callback
 
+# remove message from multiple boxes
+Imap::multiremove = (paths, callback) ->
+    async.eachSeries paths, ({path, uid}, cb) =>
+        @deleteMessageInBox path, uid, cb
```
Comment (frankrousseau) [inline server/imap/connection.coffee]: Is that possible to make a bulk delete with IMAP?

File: server/models/message.coffee [hunk 9]
```diff
@@ -541,6 +560,7 @@ Message.createFromImapMessage = (mail, box, uid, callback) ->
 
     # we store normalized versions of subject & messageID for threading
     messageID = mail.headers['message-id']
+    delete mail.messageId
```
Comment (frankrousseau) [inline server/models/message.coffee]: I read that it's more performant to do that:

```coffeescript
mail.mesesageId = undefined
```

File: server/models/message.coffee [hunk 10]
```diff
@@ -571,6 +592,8 @@ Message.createFromImapMessage = (mail, box, uid, callback) ->
                 name: att.generatedFileName
                 buffer: buffer
 
+        delete mail.attachments
```
Comment (frankrousseau) [inline server/models/message.coffee]: I read that it's more performant to do that:

```coffeescript
mail.mesesageId = undefined
```

File: client/app/components/toolbox_actions.coffee (hunk 1 above)
Comment (aenario) [inline client/app/components/toolbox_actions.coffee]: This is not related to cozydb.

It is a seprate bugfix :  in conversation you could not see some raw messages.
Exemple : you have message A in inbox and message B in sent. You display the conversation from message A and it shows message B, but the message B is not in inbox so the messageRawID was undefined.

File: server/config.coffee (hunk 2 above)
Comment (aenario) [inline server/config.coffee]: the pouchdb compatibility should be provided by cozydb (we discussed it in another PR, it still needs to be tested)

File: server/controllers/messages.coffee (hunk 3 above)
Comment (aenario) [inline server/controllers/messages.coffee]: We kept the full file in memory before and still do. We need to build the email to send it through SMTP & IMAP. Node-imap support for streams2 is not very good so until we can spend a few days making it solid, it is safer to buffer everything. 

The reason for this change is that jdb-cozy-adapter returned the mikeal/request object which act as a full stream (.on 'data') whereas cozydb return a hackish "LaterStream" which only has the .pipe and .abort method. I figure it was simpler to make this hack and remove it when we do the full streaming mentioned in the first § 

It will be cleaner to have it as a "Bufferer" class in helpers. Do you want me to make this change ?

File: server/controllers/messages.coffee (hunk 4 above)
Comment (aenario) [inline server/controllers/messages.coffee]: will do, thanks.

File: server/config.coffee (hunk 2 above)
Comment (frankrousseau) [inline server/config.coffee]: What I mean is that, the MODEL_MODULE var ensured that pouch was used in case of standalone applicaiton. So, is it something removed or cozydb handles it itself?

File: server/imap/connection.coffee (hunk 8 above)
Comment (aenario) [inline server/imap/connection.coffee]: It is possible to remove multiple message from a mailbox, but not as we are doing here a message that is in multiple mailboxes.

File: server/controllers/messages.coffee (hunk 3 above)
Comment (frankrousseau) [inline server/controllers/messages.coffee]: Don't waste too much time on it, just makes sure it works and explain everything through comments.

File: server/imap/connection.coffee (hunk 8 above)
Comment (frankrousseau) [inline server/imap/connection.coffee]: ok

File: server/models/message.coffee (hunk 10 above)
Comment (aenario) [inline server/models/message.coffee]: var a = {}
Object.keys(a) ---> Array [  ]
a.test = 'value'
Object.keys(a) ---> Array [ "test" ]
a.test = undefined
Object.keys(a) ---> Array [ "test" ]
delete a.test
Object.keys(a)   --->  Array [  ]
It is faster for "hot" objects (it prevents some interpreter optimization with hidden class), but its not exactly the same. I'd rather not risk introducing subtle bugs for a µs improvement of a once-called route.

File: server/config.coffee (hunk 2 above)
Comment (aenario) [inline server/config.coffee]: cozydb handle it itself : if there is a db or dbName in americano options, it switch to pouchdb.
This should cover the cozy-light case. For the standole version to work, we need to set db or dbName from environement variable just above (https://github.com/aenario/cozy-emails/blob/d7efc8779304fb1abeb29d50b0bb9d2661018917/server.coffee#L15)

File: server/models/message.coffee (hunk 10 above)
Comment (frankrousseau) [inline server/models/message.coffee]: Ok.

File: server/config.coffee (hunk 2 above)
Comment (frankrousseau) [inline server/config.coffee]: ok

**Comment by frankrousseau** [comment]
@clochix @aenario Can we merge it?

**Comment by aenario** [comment]
i think so

**Comment by aenario** [comment]
Let's merge it and fix bug in the next 3 days.
